alf/utils/process_coordinator.py

```diff
@@ -0,0 +1,337 @@
+        # down without losing too many cycles.
+        # The sleep duration is limited to the remaining grace duration.
+        stop_wait_secs = 0.001
+        while any(t.is_alive()
+                  for t in processes) and stop_grace_period_secs >= 0.0:
+            time.sleep(stop_wait_secs)
+            stop_grace_period_secs -= stop_wait_secs
+            stop_wait_secs = 2 * stop_wait_secs
+            # Keep the waiting period within sane bounds.
+            # The minimum value is to avoid decreasing stop_wait_secs to a value
+            # that could cause stop_grace_period_secs to remain unchanged.
+            stop_wait_secs = max(
+                min(stop_wait_secs, stop_grace_period_secs), 0.001)
+
+        # List the processes still alive after the grace period.
+        stragglers = [t.name for t in processes if t.is_alive()]
+
+        # Terminate with an exception if appropriate.
+        with self._lock:
+            self._joined.value = 1
+            self._registered_processes = set()
+            if self._exc_info_to_raise:
+                six.reraise(*self._exc_info_to_raise)
+            elif stragglers:
+                if ignore_live_processes:
+                    logging.info(
+                        "Coordinator stopped with processes still running: %s",
+                        " ".join(stragglers))
+                else:
+                    raise RuntimeError(
+                        "Coordinator stopped with processes still running: %s"
+                        % " ".join(stragglers))
+
+    @property
+    def joined(self):
+        return self._joined.value
+
+    def raise_requested_exception(self):
+        """If an exception has been passed to `request_stop`, this raises it."""
+        with self._lock:
+            if self._exc_info_to_raise:
+                six.reraise(*self._exc_info_to_raise)
+
+
+class Process(mp.Process):
+    """A coordinated process class to execute acting loops.
+    """
+
+    def __init__(self, coord, target=None, args=(), kwargs={}):
+        """Creates a process, running target in a loop, managed by coordinator.
+
+        Args:
+            coord (Coordinator): coordinator used to manage this new process.
+            target (callable): to be invoked by run() in a loop, until
+                coordinator tells the process to stop.
+            args (list): optional arguments for target callable.
+            kwargs (dict): optional keyword arguments for target callable.
+        """
+        if not isinstance(coord, Coordinator):
+            raise ValueError(
+                "'coord' argument must be a Coordinator: %s" % coord)
+        super().__init__()
+        self._coord = coord
+        # allow pass in target or overriding body
+        self._target = target or self.body
+        self._args = args
+        self._kwargs = kwargs
+        self._coord.register_process(self)
+
+    @staticmethod
+    def process(coord, target, args=(), kwargs={}):
```

Comment:
Also, I don't think that this function is necessary. It's better that p.start() won't be automatically called. In this way, the main process can manage a group of subprocesses and decide when to start.